File: 0078-subsets/0078-subsets.java
class Solution {
    List<List<Integer>> res;
    List<Integer> path;
    public List<List<Integer>> subsets(int[] nums) {
        res = new ArrayList<>();
        path = new ArrayList<>();
        backtracking(nums, 0);
        return res;
    }

    private void backtracking(int[] nums, int start) {
        res.add(new ArrayList<Integer>(path));

        for (int i = start; i < nums.length; i++) {
            path.add(nums[i]);
            backtracking(nums, i + 1);
            path.remove(path.size() - 1);
        }
    }
}

File: 0078-subsets/0078-subsets.ts
function subsets(nums: number[]): number[][] {
    const res : number[][] = [];
    const subset : number[] = [];
    
    function backtrack(start: number) : void {
        res.push([...subset]);
        
        for (let i = start; i < nums.length; i++) {
            subset.push(nums[i]);
            backtrack(i + 1);
            subset.pop();
        }
    }

    backtrack(0);
    return res;
};